File: src/core/executor.ts
import { spawn } from "node:child_process";
import type { CIJob, ExecResult } from "./models.js";

const PKG_MGR_RE = /^\s*(apt-get|apt|yum|dnf|apk)\s/;

function needsSudo(cmd: string): boolean {
  if (process.getuid?.() === 0) return false;
  return PKG_MGR_RE.test(cmd);
}

function prependSudo(cmd: string): string {
  const parts = cmd.split(/(&&|\|\|)/);
  return parts
    .map((part) => {
      const stripped = part.trim();
      if (stripped === "&&" || stripped === "||") return part;
      if (PKG_MGR_RE.test(stripped)) {
        return part.replace(stripped, `sudo ${stripped}`);
      }
      return part;
    })
    .join("");
}

function runShellCommand(
  cmd: string,
  cwd: string,
  env: NodeJS.ProcessEnv,
  timeoutMs: number,
  signal?: AbortSignal,
): Promise<{ stdout: string; exitCode: number; timedOut: boolean; cancelled: boolean }> {
  return new Promise((resolve) => {
    if (signal?.aborted) {
      resolve({ stdout: "", exitCode: -1, timedOut: false, cancelled: true });
      return;
    }

    const proc = spawn("sh", ["-c", cmd], {
      cwd,
      env,
      stdio: ["ignore", "pipe", "pipe"],
      detached: true,
    });
    const chunks: Buffer[] = [];
    let timedOut = false;
    let cancelled = false;
    let done = false;

    const kill = () => {
      const pid = proc.pid;
      if (pid === undefined) {
        proc.kill("SIGKILL");
        return;
      }
      try {
        process.kill(-pid, "SIGKILL");
      } catch {
        proc.kill("SIGKILL");
      }
    };

    const timer = setTimeout(() => {
      if (!done) {
        timedOut = true;
        kill();
      }
    }, timeoutMs);

    const onAbort = () => {
      if (!done) {
        cancelled = true;
        kill();
      }
    };
    signal?.addEventListener("abort", onAbort, { once: true });

    proc.stdout?.on("data", (chunk: Buffer) => chunks.push(chunk));
    proc.stderr?.on("data", (chunk: Buffer) => chunks.push(chunk));

    proc.on("close", (code) => {
      done = true;
      clearTimeout(timer);
      signal?.removeEventListener("abort", onAbort);
      let exitCode: number;
      if (cancelled) exitCode = -1;
      else if (timedOut) exitCode = 124;
      else exitCode = code ?? -1;
      resolve({
        stdout: Buffer.concat(chunks).toString("utf-8"),
        exitCode,
        timedOut,
        cancelled,
      });
    });

    proc.on("error", (err) => {
      done = true;
      clearTimeout(timer);
      signal?.removeEventListener("abort", onAbort);
      resolve({
        stdout: `Stitch: failed to spawn: ${err}\n`,
        exitCode: 127,
        timedOut: false,
        cancelled: false,
      });
    });
  });
}

export class LocalExecutor {
  repoRoot: string;
  timeoutSeconds: number;

  constructor(repoRoot: string, timeoutSeconds = 300) {
    this.repoRoot = repoRoot;
    this.timeoutSeconds = timeoutSeconds;
  }

  async runJob(job: CIJob, signal?: AbortSignal): Promise<ExecResult> {
    if (job.script.length === 0) {
      return {
        log: "(job has no script commands)",
        exitCode: 0,
        timedOut: false,
        durationSeconds: 0,
      };
    }

    const start = performance.now();
    const logParts: string[] = [];
    const env = { ...process.env, STITCH_RUN: "1" };
    let remaining = this.timeoutSeconds * 1000;

    for (const rawCmd of job.script) {
      if (signal?.aborted) {
        const duration = (performance.now() - start) / 1000;
        return {
          log: logParts.join(""),
          exitCode: -1,
          timedOut: false,
          cancelled: true,
          durationSeconds: duration,
        };
      }

      const cmd = needsSudo(rawCmd) ? prependSudo(rawCmd) : rawCmd;
      const cmdStart = performance.now();
      logParts.push(`$ ${cmd}\n`);

      const result = await runShellCommand(
        cmd,
        this.repoRoot,
        env,
        Math.max(100, remaining),
        signal,
      );
      logParts.push(result.stdout);
      const elapsed = performance.now() - cmdStart;
      remaining -= elapsed;

      if (result.cancelled) {
        const duration = (performance.now() - start) / 1000;
        return {
          log: logParts.join(""),
          exitCode: -1,
          timedOut: false,
          cancelled: true,
          durationSeconds: duration,
        };
      }

      if (result.timedOut) {
        logParts.push(`\nStitch: command timed out after ${this.timeoutSeconds}s\n`);
        const duration = (performance.now() - start) / 1000;
        return { log: logParts.join(""), exitCode: 124, timedOut: true, durationSeconds: duration };
      }

      if (result.exitCode !== 0) {
        const duration = (performance.now() - start) / 1000;
        return {
          log: logParts.join(""),
          exitCode: result.exitCode,
          timedOut: false,
          durationSeconds: duration,
        };
      }

      if (remaining <= 0) {
        logParts.push(`\nStitch: overall job timeout reached (${this.timeoutSeconds}s)\n`);
        const duration = (performance.now() - start) / 1000;
        return { log: logParts.join(""), exitCode: 124, timedOut: true, durationSeconds: duration };
      }
    }

    const duration = (performance.now() - start) / 1000;
    return { log: logParts.join(""), exitCode: 0, timedOut: false, durationSeconds: duration };
  }
}

File: stitch_agent/run/executor.py
"""Local job executor — run CI job scripts as subprocesses."""

from __future__ import annotations

import asyncio
import os
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path

    from stitch_agent.run.models import CIJob


@dataclass
class ExecResult:
    log: str
    exit_code: int
    timed_out: bool = False
    duration_seconds: float = 0.0


class LocalExecutor:
    """Execute CIJob script commands locally via shell subprocess."""

    def __init__(
        self, repo_root: Path, timeout_seconds: float = 300.0,
    ) -> None:
        self.repo_root = repo_root
        self.timeout_seconds = timeout_seconds

    async def run_job(self, job: CIJob) -> ExecResult:
        """Run a job's script sequentially.

        Stops on first non-zero exit. Combines stdout+stderr. Enforces an
        overall timeout across the entire job script.
        """
        if not job.script:
            return ExecResult(log="(job has no script commands)", exit_code=0)

        start = time.monotonic()
        combined_log_parts: list[str] = []
        env = dict(os.environ)
        env.setdefault("STITCH_RUN", "1")

        remaining = self.timeout_seconds
        for cmd in job.script:
            cmd_start = time.monotonic()
            combined_log_parts.append(f"$ {cmd}\n")
            try:
                proc = await asyncio.create_subprocess_shell(
                    cmd,
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.STDOUT,
                    cwd=str(self.repo_root),
                    env=env,
                )
            except Exception as exc:  # spawn failure (shell not found etc.)
                combined_log_parts.append(f"Stitch: failed to spawn: {exc}\n")
                duration = time.monotonic() - start
                return ExecResult(
                    log="".join(combined_log_parts),
                    exit_code=127,
                    timed_out=False,
                    duration_seconds=duration,
                )

            try:
                stdout_bytes, _ = await asyncio.wait_for(
                    proc.communicate(), timeout=max(0.1, remaining),
                )
            except TimeoutError:
                with _suppress():
                    proc.kill()
                with _suppress():
                    await proc.wait()
                combined_log_parts.append(
                    f"\nStitch: command timed out after {self.timeout_seconds}s\n"
                )
                duration = time.monotonic() - start
                return ExecResult(
                    log="".join(combined_log_parts),
                    exit_code=124,
                    timed_out=True,
                    duration_seconds=duration,
                )

            combined_log_parts.append(
                (stdout_bytes or b"").decode("utf-8", errors="replace")
            )
            exit_code = proc.returncode if proc.returncode is not None else -1
            remaining -= time.monotonic() - cmd_start
            if exit_code != 0:
                duration = time.monotonic() - start
                return ExecResult(
                    log="".join(combined_log_parts),
                    exit_code=exit_code,
                    timed_out=False,
                    duration_seconds=duration,
                )
            if remaining <= 0:
                combined_log_parts.append(
                    f"\nStitch: overall job timeout reached ({self.timeout_seconds}s)\n"
                )
                duration = time.monotonic() - start
                return ExecResult(
                    log="".join(combined_log_parts),
                    exit_code=124,
                    timed_out=True,
                    duration_seconds=duration,
                )

        duration = time.monotonic() - start
        return ExecResult(
            log="".join(combined_log_parts),
            exit_code=0,
            timed_out=False,
            duration_seconds=duration,
        )


class _suppress:
    def __enter__(self) -> None:
        return None

    def __exit__(self, exc_type: object, exc: object, tb: object) -> bool:
        return True
